### PR TITLE
Implement GccAsmStatement for gcc-style asm statements

### PR DESCRIPTION
**Pull request by ibuclaw**
The AST itself is only used by GDC.

Parser and semantic analysis has been merged into a single function, and put in its own module as per https://github.com/dlang/dmd/pull/7562#issuecomment-360684196.

Lots more documentation could be added to the new functions that implement gcc asm.

Closes #7562.

### CODE REVIEW

**Comment by ibuclaw** [comment]
As title update suggests, renamed class to GccAsmStatement to avoid ambiguity.

**Review by JinShil — APPROVED** [review]

File: src/dmd/iasmgcc.d [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+        s.insn = (cast(CompileExp)s.insn).e1;
+    s.insn = semanticString(sc, s.insn, "asm instruction template");
+
+    if (s.labels && s.outputargs)
+        s.error("extended asm statements with labels cannot have output constraints");
+
+    // Analyse all input and output operands.
+    if (s.args)
+    {
+        for (size_t i = 0; i < s.args.dim; i++)
```
Comment (wilzbach) [inline src/dmd/iasmgcc.d]: Could be a foreach

**Review by wilzbach — APPROVED** [review]
Reviewing this commit by commit was really straight-forward. Thank you!

File: src/dmd/iasmgcc.d (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/iasmgcc.d]: If you mean `foreach (i; 0 .. s.args.dim)`, then ok.

**Comment by ibuclaw** [comment]
Rebased; tweaked documented grammar so that there is no distinction between input and output operands; use foreach in a few places.

**Comment by ibuclaw** [comment]
@WalterBright - This is good to review I guess.

**Comment by jacob-carlborg** [comment]
Can tests be added?

**Comment by ibuclaw** [comment]
@jacob-carlborg unittests mimicking gdc? I guess so...

**Comment by jacob-carlborg** [comment]
> jacob-carlborg unittests mimicking gdc? I guess so..

I had a few ideas:

1. Use more of an integration/unit test style approach instead of the current end-to-end
2. Tests that shows what happens if DMD is trying to use this new statement
3. Can we have tests that are only executed by GDC? We are running GDC using SemaphoreCI, as far as I know

Unless somebody has a better idea.

**Comment by ibuclaw** [comment]
>  Use more of an integration/unit test style approach instead of the current end-to-end

See latest commit.

> Tests that shows what happens if DMD is trying to use this new statement

DMD will always treat it like a dmd asm statement, so will error in same way as before.

> Can we have tests that are only executed by GDC? We are running GDC using SemaphoreCI, as far as I know

That doesn't really help at all if we make changes to it, as for proper testing you should build GDC from source.  You should probably have some faith in me that I vet my own compiler. :-)

**Comment by jacob-carlborg** [comment]
> That doesn't really help at all if we make changes to it, as for proper testing you should build GDC from source. You should probably have some faith in me that I vet my own compiler

Yes, but you're making changes to a project without adding tests to the project. But now I see that you've added tests :+1:.

**Review by PetarKirov — APPROVED** [review]
Nice work on the code and git commit separation!

**Comment by wilzbach** [comment]
SemaphoreCI failure is due to it checking out the wrong branch. This should be fixed by https://github.com/dlang/dmd/pull/8560 (once it's merged into `stable` and merged to `master`).

**Comment by ibuclaw** [comment]
:-O

**Pull request merged**